### PR TITLE
feat: support automatic json body for post requests

### LINKED ISSUE #4: Supporting JSON body
It would be great for POST request to support body as an object.
If this is the case, ohmyfetch will add the correct header and stringify the body.

### PR DESCRIPTION
**Pull request by danielroe**
closes #4

### CODE REVIEW

**Review by Atinux — APPROVED** [review]

File: src/fetch.ts [hunk 1]
```diff
@@ -31,8 +32,21 @@ export function createFetch ({ fetch }: CreateFetchOptions): $Fetch {
...
+          opts.body = JSON.stringify(opts.body)
+          opts.headers = opts.headers || {}
+          if (Array.isArray(opts.headers)) {
+            opts.headers.push(['content-type', 'application/json'])
+          } else if ('set' in opts.headers) {
+            ;(opts.headers as Headers).set('content-type', 'application/json')
+          } else {
+            opts.headers['content-type'] = 'application/json'
+          }
+        } catch {}
```
Comment (pi0) [inline src/fetch.ts:46]: try/catch scope is wider than necessary (i think purpose was to handle JSON.stringify issues) but i think for that we may either keep original error or fallback to opts.body

File: src/fetch.ts [hunk 2]
```diff
@@ -31,8 +32,21 @@ export function createFetch ({ fetch }: CreateFetchOptions): $Fetch {
       if (opts.params) {
         request = withQuery(request, opts.params)
       }
+      if (opts.body && typeof opts.body === 'object' && ['patch', 'post', 'put'].includes(opts.method?.toLowerCase() || '')) {
```
Comment (pi0) [inline src/fetch.ts:35]: Body can be also of type `FormData` (typeof === object). We should ensure proto is pure object

Comment (pi0) [inline src/fetch.ts:35]: Also i would create an array outside for keeping methods that accept post

File: src/fetch.ts [hunk 3]
```diff
@@ -31,8 +32,21 @@ export function createFetch ({ fetch }: CreateFetchOptions): $Fetch {
...
+        try {
+          opts.body = JSON.stringify(opts.body)
+          opts.headers = opts.headers || {}
+          if (Array.isArray(opts.headers)) {
+            opts.headers.push(['content-type', 'application/json'])
```
Comment (pi0) [inline src/fetch.ts:40]: Note: In array format, we might add duplicate content-type.

File: src/fetch.ts [hunk 4]
```diff
@@ -32,18 +34,21 @@ export function createFetch ({ fetch }: CreateFetchOptions): $Fetch {
...
-          if (Array.isArray(opts.headers)) {
-            opts.headers.push(['content-type', 'application/json'])
-          } else if ('set' in opts.headers) {
-            ;(opts.headers as Headers).set('content-type', 'application/json')
+      if (opts.body && opts.body.toString() === '[object Object]' && payloadMethods.includes(opts.method?.toLowerCase() || '')) {
+        opts.body = JSON.stringify(opts.body)
+        opts.headers = opts.headers || {}
+        if (Array.isArray(opts.headers)) {
+          const index = opts.headers.findIndex(([header]) => header.toLowerCase() === 'content-type')
+          if (index >= 0) {
```
Comment (pi0) [inline src/fetch.ts:42]: - Can't we inline with `headers.find`?
- Lowercase check is not happening for object mode

With currently increased logic, i would refactor header setting to a utility function

File: src/fetch.ts [hunk 5]
```diff
@@ -22,6 +25,28 @@ export interface $Fetch {
...
+    const existingHeader = options.headers.find(([header]) => header.toLowerCase() === key)
+    if (existingHeader) {
+      existingHeader[1] = value
+    } else {
+      options.headers.push([key, value])
+    }
+  } else if ('set' in options.headers) {
+    ;(options.headers as Headers).set(key, value)
+  } else {
+    const existingHeader = Object.keys(options.headers).find(header => header.toLowerCase() === key)
```
Comment (pi0) [inline src/fetch.ts:41]: By early returning on `set in` case, we can dedup this line

**Comment by pi0** [comment]
It think would be nice mentioning this feature in readme :)

File: src/fetch.ts (hunk 5 above)
Comment (danielroe) [reply src/fetch.ts:41]: not quite sure how to dedupe it - though it feels very close with the array/object syntax. how were you thinking?

Comment (pi0) [reply src/fetch.ts:41]: By early return in first if case we can move variable to outer scope :)

**Pull request merged**